jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassResolver.java

```diff
@@ -0,0 +1,136 @@
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.frontend;
+
+import com.google.cloud.tools.jib.configuration.BuildConfiguration;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Verify;
+import javax.annotation.Nullable;
+
+/** Infers the main class in an application. */
+public class MainClassResolver {
+
+  /**
+   * If {@code mainClass} is {@code null}, tries to infer main class in this order:
+   *
+   * <ul>
+   *   <li>1. Looks in a {@code jar} plugin provided by {@code projectProperties} ({@code
+   *       maven-jar-plugin} for maven or {@code jar} task for gradle).
+   *   <li>2. Searches for a class defined with a main method.
+   * </ul>
+   *
+   * <p>Warns if main class provided by {@code projectProperties} is not valid, or throws an error
+   * if no valid main class is found.
+   *
+   * @param mainClass the explicitly configured main class ({@code null} if not configured).
+   * @param projectProperties properties containing plugin information and help messages.
+   * @return the name of the main class to be used for the container entrypoint.
+   * @throws MainClassInferenceException if no valid main class is configured or discovered.
+   */
+  public static String resolveMainClass(
+      @Nullable String mainClass, ProjectProperties projectProperties)
+      throws MainClassInferenceException {
+    // If mainClass is null, try to find via projectProperties.
+    if (mainClass == null) {
+      mainClass = getMainClassFromJar(projectProperties);
+    }
+
+    // If mainClass is still null, try to search in class files.
+    if (mainClass == null) {
+      mainClass = findMainClassInClassFiles(projectProperties, true);
+
+    } else if (!BuildConfiguration.isValidJavaClass(mainClass)) {
+      // If mainClass found in projectProperties is not valid, try to search in class files, but
+      // don't error if not found in class files.
+      String mainClassInClassFiles = findMainClassInClassFiles(projectProperties, false);
+      if (mainClassInClassFiles != null) {
+        mainClass = mainClassInClassFiles;
+      }
+    }
+
+    Preconditions.checkNotNull(mainClass);
+    if (!BuildConfiguration.isValidJavaClass(mainClass)) {
+      projectProperties.getLogger().warn("'mainClass' is not a valid Java class : " + mainClass);
+    }
+
+    return mainClass;
+  }
+
+  @Nullable
+  private static String getMainClassFromJar(ProjectProperties projectProperties) {
+    projectProperties
+        .getLogger()
+        .info(
+            "Searching for main class... Add a 'mainClass' configuration to '"
+                + projectProperties.getPluginName()
+                + "' to improve build speed.");
+    return projectProperties.getMainClassFromJar();
+  }
+
+  @Nullable
+  private static String findMainClassInClassFiles(
+      ProjectProperties projectProperties, boolean shouldError) throws MainClassInferenceException {
```

Comment:
Making the method throw or suppress exceptions based on an argument seems weird to me. I'm thinking it's better to remove `shouldError` and have caller to catch the exception if they really want to ignore it.